src/core/execute.c

```diff
@@ -2189,6 +2189,9 @@ static int setup_keyring(
         if (context->keyring_mode == EXEC_KEYRING_INHERIT)
                 return 0;
 
+        if (have_effective_cap(CAP_SYS_ADMIN) > 0 && (uid_is_valid(uid) || gid_is_valid(gid)))
```

Comment:
uh, sorry for my comments earlier, this needs to check "== 0" rather than "> 0" of course. We want to exit early if we don't have CAP_SYS_ADMIN after all, but uid/gid are configured.

Also, could you add a brief comment explaining why this check is done?

Reply:
yeah, it did feel wrong. does `have_effective_cap()` return boolean? thus this should be `!have_effective_cap(CAP_SYS_ADMIN)`. Comments yes would be good. Others have also mentioned that `log_open(...)` or some such is missing, thus all the `log_*` functions are ineffective in this function. Should this be fixed too? It seems like `child_exec` closes log with `log_forget_fds()` thus things like `setup_keyring()` should reopen the log themself? or should it take error_message pointer and populate that with useful things?

Reply:
it returns < 0 if we can't determine whether the cap is around. It returns 0 if it isn't and > 0 if it is.

I figure in this case checking for == 0, so that in case we have the cap or in case we can't figure out whether we do we execute the rest of the function and otherwise exit early